### ISSUE TITLE
lighttable sluggish when scrolling down/up

### ISSUE THREAD

**Issue by TurboGit** [darktable-org/darktable#15421]
Since some time I have found that the lighttable is sluggish and mostly when many images are displayed in a single row.

To reproduce:
1. Select a collection with lot of images, say more that 500
2. Set the lighttable with 20 pcitures in a raw
3. Set display to top of collection, that is first image in the collection being displayed on top-left corner of lighttable
4. Do a single long/big scroll

See that there is some delay ~2 seconds on my computer before the lighttable refresh.

Note that the same sluggishness happens if one try to drag the scrollbar. 

We have been discussing this with @AlicVB and we found out that this is because the `_event_scroll()` routine is receiving lot of events (1 per row to move) and so that many call to `_move` is done.

I'm almost sure this was not the case some time ago, at least just after the full reimplementation of lighttable by @AlicVB .

This may be due to smooth scrolling or missing event compression.

@AlicVB : Am I missing something in this summary?

@dterrahe : As our Gtk+ expert, any idea?

**Comment by jenshannoschwalm** [comment]
Might that be related to "we don't use double buffers for writing to main gtk image any more"?

**Comment by AlicVB** [comment]
> Might that be related to "we don't use double buffers for writing to main gtk image any more"?

not sure : in my test, it seems that the images itself are not drawn...

To test that : be sure to empty the thumb cache (and maybe disable your reset work on background loading). Then scroll bottom a lot. After some delay, you will arrive at the right position. now scroll up a little, and see that the thumbnail are not cached... iiuc that proves that we don't have drawn anything on screen...

in my understanding, we are doing all the image position/size computing, the widget creation, etc, but once this is done the next scroll event is handled. before any draw event is fired by gtk... (thumbnails are load inside the draw event)

**Comment by dterrahe** [comment]
> we are doing all the image position/size computing, the widget creation, etc, but once this is done the next scroll event is handled. before any draw event is fired by gtk... (thumbnails are load inside the draw event)

If new widgets are created and added to a container _during_ a draw event, they will only be realised/mapped/sized/positioned before the _next_ draw event cycle. If they are then replaced again before they get a chance to receive a draw event themselves, they will never appear.

For libs with delayed `gui_update` I found a workaround in `dt_gui_widget_reallocate_now` that forces the layout step during the draw. This hack may not work if the effort is too "heavy"; draw event handlers should not take up multiple cycles or probably bad things will happen.

I'm not saying that you have the same problem here, but it could be.

> Might that be related to "we don't use double buffers for writing to main gtk image any more"?

No. At least I can't imagine a way it would be related. Gtk is _still_ doing the double buffering for us; we used to do triple buffering before, for no good reason and with no visible result besides a possible slowdown. Not saying that I couldn't have broken something with far reaching side effects, just that the drouple buffering would be a more surprising reason for breakage to me than just about anything else.

**Comment by AlicVB** [comment]
iirc, what we are doing in the image widget draw event is "only" : 
- loading/creating the thumbnail image
- drawing this image in the widget

no widget creation...

I really think that what is eating time here is "only" all the rest of the work (widget creation, placement, metadata reading, etc...) The problem comes from the fact that if you do a "big" scroll in one time (say 50 lines down) the code is treating each line move one by one...

**Comment by TurboGit** [comment]
@AlicVB : While looking a bit at the code I've seen (but not related to this):

Two:

```
    GList *tl = g_list_find_custom(table->list,
                                   GINT_TO_POINTER(id), _list_compare_by_imgid);
```

This list traversal done in `dt_thumbtable_full_redraw` is really not optimal as done for every image on screen and the list can be quite long. As part of this issue we might want to fix this by using a hash table.

Also, in `_event_scroll` we are not taking delta into account when moving lighttable line up or down if I read the code correctly. We might want to fix this too.

Again, this is not the perf issue we are discussing, just taking notes.

**Comment by AlicVB** [comment]
> Also, in `_event_scroll` we are not taking delta into account when moving lighttable line up or down if I read the code correctly. We might want to fix this too.

iirc, this is because delta can only be 1.0 or -1.0...



> This list traversal done in `dt_thumbtable_full_redraw` is really not optimal as done for every image on screen and the list can be quite long. As part of this issue we might want to fix this by using a hash table.

Yes : I'm sure there's lot of small improvement like that possible in this fct. But as you say, that won't solve the root issue...

As I see this, a solution would be to "compress" somehow gtk scroll events, in order to get 1 unique event for big scroll (with a big delta, of course).
@dterrahe : do you think it's possible directly with some gtk magics ?
Or maybe we can do it manually with a very small timeout in the scroll event ?

**Comment by TurboGit** [comment]
> iirc, this is because delta can only be 1.0 or -1.0...

I've seen some delta up to 3 and it seems that delta is taken into account for filmstrip.

**Comment by TurboGit** [comment]
> As I see this, a solution would be to "compress" somehow gtk scroll events, in order to get 1 unique event for big scroll

This will help for big scroll, but do a simple scroll and you'll still notice some delay before the lighttable updates. Also this seems related to the number of thumbs on a line (so the number of thumbs displayed)... But at the end for Gtk that doesn't mean moving more pixels up or down. So I'm not so sure about the real issue, we would need lot of timing on this part of code to pin point the real culprit.

**Comment by AlicVB** [comment]
ok, so after a lot of printf... here are some more information : 
- `dt_thumbtable_full_redraw` is not the culprit for "big" scroll, as it happen only once all the event have been handled.
- what seems to take the more time is `_thumbs_load_needed` and `_thumbs_remove_unneeded` called by the `_move` fct
- inside `_thumbs_load_needed`, the time is spent equally in the widget creation and the widget addition to the container. This explain why it's more visible with lot of images per row...

Starting from here, I see multiple solutions : 
1. move and reaffect widgets instead of deleting/recreating them. Although this seems quite simple, it is not, especially to take account of all the special cases we have (1 image only, zoomable, full preview, filmstrip, etc)
2. find a solution to "compress" the scroll events and do everything in 1 pass only...
3. both solutions (that would be the best of course)

All in one : that seems far from trivial and will need lot of work ;)

**Comment by dterrahe** [comment]
> iirc, what we are doing in the image widget draw event is "only"

I was referring to the tumb _table_ draw event handler `_event_draw` in thumbtable.c, not the thumb _nail_ draw event.

> * `dt_thumbtable_full_redraw` [...] happen only once all the event have been handled

Why does it need to happen at all? From the comment "reload all thumbs from scratch" it seems this should not be needed each scroll?

> @dterrahe : do you think it's possible directly with some gtk magics ?

I don't know; too much interacting complicated stuff is going on here to think there's a simple solution. If a scroll event handler takes too long to process (more than a drawing cycle) the next scroll event it receives should be compressed/aggregated. One should take care not to quickly process those events but then queue the actual handling elsewhere because that will create a backlog. No idea if that is happening here.

It does not _seem_ to me that a large number of individual mouse scroll events being dealt with individually is the problem here; when I simply click on the scroll bar (say, in the middle if currently it is at the top) it takes a couple of seconds for the updated grid of thumbnails to start being shown and the scroll bar then jumps to its new position in one step. Presumably this is just one call to `_move`, so that simply takes that long.

> * inside `_thumbs_load_needed`, the time is spent equally in the widget creation and the widget addition to the container

Is it the actual gtk widget handling or simply all the database access to get the info for each thumbnail? I suspect the latter. But there _are_ a lot of widgets involved. Switching to use standard gtk widgets for the layouting rather than doing it yourself in a large GtkDrawingArea draw event (or even in one per thumb) makes things simpler maybe, but also costs a hefty price.

This area is notoriously complicated to get right though, so gtk introduced a new way of dealing with a partially visible list by using a "fixed" set of widgets and connecting them to different items depending on scrolling position. See for example https://docs.gtk.org/gtk4/class.ListItemFactory.html for some info. Not so easy to do for the filemanager view...

**Comment by AlicVB** [comment]
Thanks @dterrahe for your comments !



> when I simply click on the scroll bar (say, in the middle if currently it is at the top) it takes a couple of seconds for the updated grid of thumbnails to start being shown and the scroll bar then jumps to its new position in one step. Presumably this is just one call to `_move`, so that simply takes that long.

this is somehow "fixed" in PR #15428 (at least the speedup is quite visible here), even if there's still room for enhancement...
